### PR TITLE
ill: borrowing request

### LINKED ISSUE #328: borrowing request form: requested changes
this is a collective issue which could require changes in invenio-app-ils and react-invenio-app-ils

- [x] costs shouldn't be required as they are not known when creating ILL
- [x] add validation

### PR DESCRIPTION
**Pull request by jrcastro2**
* Adds vocabulary for cancel_reason
* Adds valdidation and test
* Costs fields are not required
* closes https://github.com/CERNDocumentServer/cds-ils/issues/328

### CODE REVIEW

File: invenio_app_ils/ill/api.py [hunk 1]
```diff
@@ -88,11 +91,57 @@ def delete(self, **kwargs):
...
+        self.ensure_document_exists(document_pid)
+        self.ensure_library_exists(library_pid)
+
+
 class BorrowingRequest(IlsRecord):
     """ILL borrowing request class."""
 
     _pid_type = BORROWING_REQUEST_PID_TYPE
     _schema = "ill_borrowing_requests/borrowing_request-v1.0.0.json"
+    _validator = IllValidator()
```
Comment (ntarocco) [inline invenio_app_ils/ill/api.py:153]: Thank you! 👍🏼

File: invenio_app_ils/vocabularies/data/ill_cancel_reason.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+[
+
+  {
+    "type": "ill_cancel_reason",
+    "key": "USER_CANCEL",
```
Comment (ntarocco) [inline invenio_app_ils/vocabularies/data/ill_cancel_reason.json]: I am not sure about these values... were they copied from the document request ones or provided by the library?

**Review by KonstantinaStoikou — APPROVED** [review]

File: invenio_app_ils/vocabularies/data/ill_cancel_reason.json (hunk 2 above)
Comment (jrcastro2) [reply invenio_app_ils/vocabularies/data/ill_cancel_reason.json]: They were copied from the document request ones, as I understood it from the [issue](https://github.com/CERNDocumentServer/cds-ils/issues/328).

Comment (ntarocco) [reply invenio_app_ils/vocabularies/data/ill_cancel_reason.json]: Ok I see, for me they do not make much sense in this case. For example, a user cannot cancel a borrowing request.

I would have something more generic here, and then override in CDS-ILS and ask our library what are the values that they need.

Comment (jrcastro2) [reply invenio_app_ils/vocabularies/data/ill_cancel_reason.json]: Just for the record, librarians agreed to keep cancel_reason as a text input.

File: invenio_app_ils/ill/api.py [hunk 3]
```diff
@@ -88,11 +91,57 @@ def delete(self, **kwargs):
...
+            )
+        if cancel_reason and not status == "CANCELLED":
+            raise ILLError(
+                "If you select a cancel reason you need to select"
+                " \"Cancelled\" in the state"
+            )
+
+    def ensure_document_exists(self, document_pid):
+        """Ensure document exists or raise."""
+        from invenio_app_ils.documents.api import Document
```
Comment (ntarocco) [inline invenio_app_ils/ill/api.py]: Please do not hardcode classes import whenever possible. In this case it is not very relevant, but we tend to copy/paste around...
```suggestion
        Document = current_app_ils.document_record_cls
```

File: invenio_app_ils/items/api.py [hunk 4]
```diff
@@ -56,7 +53,7 @@ def ensure_document_exists(self, document_pid):
         try:
             Document.get_record_by_pid(document_pid)
         except PIDDoesNotExistError:
-            raise ItemDocumentNotFoundError(document_pid)
+            raise DocumentNotFoundError(document_pid)
```
Comment (ntarocco) [inline invenio_app_ils/items/api.py:56]: it would be nice to fix the import here too if possible

**Review by ntarocco — APPROVED** [review]

File: invenio_app_ils/errors.py [hunk 5]
```diff
@@ -213,6 +213,17 @@ def __init__(self, document_pid, **kwargs):
         self.description = self.description.format(document_pid)
 
 
+class LibraryNotFoundError(IlsException):
```
Comment (ntarocco) [inline invenio_app_ils/errors.py]: Sorry, can you move this inside the ILL package/module? we tried to keep ILL/ACQ isolated and independent as much as possible. Thanks!
```suggestion
class LibraryNotFoundError(ILLError):
```

File: invenio_app_ils/ill/api.py [hunk 6]
```diff
@@ -88,11 +91,57 @@ def delete(self, **kwargs):
...
+
+        try:
+            Document.get_record_by_pid(document_pid)
+        except PIDDoesNotExistError:
+            raise DocumentNotFoundError(document_pid)
+
+    def ensure_library_exists(self, library_pid):
+        """Ensure library exists or raise."""
+        try:
+            Library.get_record_by_pid(library_pid)
```
Comment (ntarocco) [inline invenio_app_ils/ill/api.py:128]:
```suggestion
            Library = current_ils_ill.library_record_cls
            Library.get_record_by_pid(library_pid)
```

File: invenio_app_ils/errors.py [hunk 7]
```diff
@@ -213,6 +213,17 @@ def __init__(self, document_pid, **kwargs):
...
+    """Raised when a library could not be found."""
+
+    description = "Library PID '{}' was not found"
+
+    def __init__(self, document_pid, **kwargs):
```
Comment (ntarocco) [inline invenio_app_ils/errors.py]:
```suggestion
    def __init__(self, library_pid, **kwargs):
```

File: invenio_app_ils/errors.py [hunk 8]
```diff
@@ -213,6 +213,17 @@ def __init__(self, document_pid, **kwargs):
...
+        """Initialize exception."""
+        super().__init__(**kwargs)
+        self.description = self.description.format(document_pid)
```
Comment (ntarocco) [inline invenio_app_ils/errors.py]:
```suggestion
        self.description = self.description.format(library_pid)
```

File: invenio_app_ils/ill/api.py [hunk 9]
```diff
@@ -88,11 +91,57 @@ def delete(self, **kwargs):
...
+        except PIDDoesNotExistError:
+            raise LibraryNotFoundError(library_pid)
+
+    def validate(self, record, **kwargs):
+        """Validate record before create and commit."""
+        super().validate(record, **kwargs)
+
+        status = record.get("status", None)
```
Comment (ntarocco) [inline invenio_app_ils/ill/api.py]: Mandatory field:
```suggestion
        status = record["status"]
```
can you please do the same from document and library?

File: invenio_app_ils/ill/api.py [hunk 10]
```diff
@@ -88,11 +91,57 @@ def delete(self, **kwargs):
...
+        cancel_reason = record.get("cancel_reason", None)
+        document_pid = record.get("document_pid", None)
+        library_pid = record.get("library_pid", None)
```
Comment (ntarocco) [inline invenio_app_ils/ill/api.py]: can you also check that the patron exists?
`patron_pid = record["patron_pid"]`

```
    def ensure_patron_exists(self, patron_pid):
        """Ensure patron exists or raise."""
        try:
            current_app_ils.patron_cls.get_patron(patron_pid)
        except PatronNotFoundError:
            raise PatronNotFoundError(patron_pid)

```

**Review by ntarocco — APPROVED** [review]

**Pull request merged**